system/jlib/jptree.cpp

```diff
@@ -8677,6 +8687,24 @@ void removeConfigUpdateHook(unsigned notifyFuncId)
 }
 #endif // __linux__
 
+void CConfigUpdateHook::clear()
+{
+    unsigned id = configCBId.exchange((unsigned)-1);
+    if ((unsigned)-1 != id)
+        removeConfigUpdateHook(id);
+}
+
+void CConfigUpdateHook::installOnce(ConfigUpdateFunc callbackFunc, bool callWhenInstalled)
+{
+    unsigned uninitialized = (unsigned)-1;
+    if (configCBId.compare_exchange_strong(uninitialized, 0))
```

Comment:
One slight problem with this function is that it can return before the callbackfunc has been called if it can be called concurrently on multiple threads.  That probably isn't what is expected/desired.
I don't think it will cause any issue with the current code (TpWrapper is the only potential).  Adding a critsec to the class is a resonable amount of overhead.  Sharing one between all instances is robably ok, since only used if not already installed.

(It would be slightly more efficient to read the value first and only perform the compare_exchange_strong if equal -1.)

Reply:
right, yes there's a window it can return before called, I agree that's a problem.
I don't see an issue with adding a CS to this class to protect against that though.